core/src/main/scala/kafka/log/Log.scala

```diff
@@ -355,11 +355,13 @@ class Log(@volatile var dir: File,
   private def recoverSegment(segment: LogSegment, leaderEpochCache: Option[LeaderEpochFileCache] = None): Int = lock synchronized {
     val stateManager = new ProducerStateManager(topicPartition, dir, maxProducerIdExpirationMs)
     stateManager.truncateAndReload(logStartOffset, segment.baseOffset, time.milliseconds)
-    logSegments(stateManager.mapEndOffset, segment.baseOffset).foreach { segment =>
-      val startOffset = math.max(segment.baseOffset, stateManager.mapEndOffset)
-      val fetchDataInfo = segment.read(startOffset, None, Int.MaxValue)
-      if (fetchDataInfo != null)
-        loadProducersFromLog(stateManager, fetchDataInfo.records)
+    if (stateManager.mapEndOffset < segment.baseOffset) {
```

Comment:
The basic intent of this logic is to ensure that we have consistent producer state up to the base offset of the segment we're recovering before we begin the actual recovery. In this case, the log start offset is somewhere above the base offset of this segment, which makes the call to `truncateAndReload` above a little strange (i.e. the end offset is smaller than the start offset). This seems to work, but feels brittle. Since we do not actually do partial segment recovery, I wonder if we could do something like this:
```scala
val recoveryStartOffset = math.min(logStartOffset, segment.baseOffset)
stateManager.truncateAndReload(recoveryStartOffset, segment.baseOffset, time.milliseconds)
```
Then the map end offset is always no greater than the segment base offset. What do you think?

Reply:
Thanks for the context. Given we want to have consistent producer state up to the base offset of the segment we are recovering, and truncateAndReload will set the map end offset properly in this case, this approach seems better. I also tested it just now, and it works (fixes the original bug). I will update the PR with your suggestion.